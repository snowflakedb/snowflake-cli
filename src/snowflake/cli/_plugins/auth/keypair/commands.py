from pathlib import Path

import typer
from snowflake.cli._plugins.auth.keypair.manager import AuthManager, PublicKeyProperty
from snowflake.cli.api.commands.flags import SecretTypeParser
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.output.types import (
    CollectionResult,
    CommandResult,
    MessageResult,
    SingleQueryResult,
)
from snowflake.cli.api.secret import SecretType
from snowflake.cli.api.secure_path import SecurePath

app = SnowTyperFactory(
    name="keypair",
    help="Manages authentication.",
)


KEY_PAIR_DEFAULT_PATH = "~/.ssh"


def _show_connection_name_prompt(ctx: typer.Context, value: str):
    for param in ctx.command.params:
        if param.name == "connection_name":
            if value:
                param.prompt = "Enter connection name"
            break
    return value


_new_connection_option = typer.Option(
    True,
    help="Create a new connection.",
    prompt="Create a new connection?",
    callback=_show_connection_name_prompt,
    show_default=False,
    hidden=True,
)


_connection_name_option = typer.Option(
    None,
    help="The new connection name.",
    prompt=False,
    show_default=False,
    hidden=True,
)


_key_length_option = typer.Option(
    2048,
    "--key-length",
    help="The RSA key length.",
    prompt="Enter key length",
)


_output_path_option = typer.Option(
    KEY_PAIR_DEFAULT_PATH,
    "--output-path",
    help="The output path for private and public keys",
    prompt="Enter output path",
)


_private_key_passphrase_option = typer.Option(
    "",
    "--private-key-passphrase",
    help="The private key passphrase.",
    click_type=SecretTypeParser(),
    prompt="Enter private key passphrase",
    hide_input=True,
    show_default=False,
)


@app.command("setup", requires_connection=True)
def setup(
    new_connection: bool = _new_connection_option,
    connection_name: str = _connection_name_option,
    key_length: int = _key_length_option,
    output_path: Path = _output_path_option,
    private_key_passphrase: SecretType = _private_key_passphrase_option,
    **options,
):
    """
    Generates the key pair, sets the public key for the user in Snowflake and creates or updates the connection.
    """
    AuthManager().setup(
        connection_name=connection_name,
        key_length=key_length,
        output_path=SecurePath(output_path),
        private_key_passphrase=private_key_passphrase,
    )
    return MessageResult(f"Setup completed.")


@app.command("rotate", requires_connection=True)
def rotate(
    new_connection: bool = _new_connection_option,
    connection_name: str = _connection_name_option,
    key_length: int = _key_length_option,
    output_path: Path = _output_path_option,
    private_key_passphrase: SecretType = _private_key_passphrase_option,
    **options,
):
    """
    Rotates the key for the connection. Generates the key pair, sets the public key for the user in Snowflake and creates or updates the connection.
    """
    AuthManager().rotate(
        connection_name=connection_name,
        key_length=key_length,
        output_path=SecurePath(output_path),
        private_key_passphrase=private_key_passphrase,
    )
    return MessageResult(f"Rotate completed.")


@app.command("list", requires_connection=True)
def list_keys(**options) -> CommandResult:
    """
    Lists the public keys set for the user.
    """
    return CollectionResult(AuthManager().list_keys())


@app.command("remove", requires_connection=True)
def remove(
    public_key_property: PublicKeyProperty = typer.Option(
        ...,
        "--key-id",
        help=f"local path to template directory or URL to git repository with templates.",
        show_default=False,
    ),
    **options,
):
    """
    Removes the public key for the user.
    """
    return SingleQueryResult(
        AuthManager().remove_public_key(public_key_property=public_key_property)
    )


@app.command("status", requires_connection=True)
def status(**options):
    """
    Verifies the key pair configuration and tests the connection.
    """
    AuthManager().status()
    return MessageResult("Status check completed.")
