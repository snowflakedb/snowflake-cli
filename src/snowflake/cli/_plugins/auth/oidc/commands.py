# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import typer
from snowflake.cli._app.auth.oidc_providers import (
    OidcProviderTypeWithAuto,
)
from snowflake.cli._plugins.auth.oidc.manager import OidcManager
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.output.types import MessageResult

app = SnowTyperFactory(
    name="oidc",
    help="Manages OIDC authentication.",
)


AutoProviderTypeOption = typer.Option(
    OidcProviderTypeWithAuto.AUTO.value,
    "--type",
    help=f"Type of OIDC provider to use",
    show_default=False,
)


@app.command("read-token", requires_connection=False)
def read_token(
    _type: OidcProviderTypeWithAuto = AutoProviderTypeOption,
    **options,
):
    """
    Reads OIDC token based on the specified type.
    Use 'auto' to auto-detect available providers.
    """
    result = OidcManager().read_token(provider_type=_type)
    return MessageResult(result)
