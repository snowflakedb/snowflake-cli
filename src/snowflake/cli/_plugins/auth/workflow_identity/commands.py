# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import typer
from snowflake.cli._plugins.auth.workflow_identity.manager import (
    WorkflowIdentityManager,
)
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.output.types import (
    CommandResult,
    MessageResult,
    SingleQueryResult,
)

# Main workflow-identity app
app = SnowTyperFactory(
    name="workflow-identity",
    help="Manages GitHub workflow identity federation authentication.",
)


def _setup_command(
    github_repository: str = typer.Option(
        ...,
        "--github-repository",
        help="GitHub repository in format 'owner/repo'",
        prompt="Enter GitHub repository (owner/repo)",
    ),
    **options,
):
    """
    Sets up GitHub workflow identity federation for authentication.
    """
    WorkflowIdentityManager().setup(github_repository=github_repository)
    return MessageResult("GitHub workflow identity federation setup completed.")


def _status_command(**options) -> CommandResult:
    """
    Shows the status of GitHub workflow identity federation configuration.
    """
    result = WorkflowIdentityManager().status()
    return MessageResult(f"GitHub workflow identity federation status: {result}")


def _remove_command(**options) -> CommandResult:
    """
    Removes the GitHub workflow identity federation configuration.
    """
    return SingleQueryResult(WorkflowIdentityManager().remove())


# Register commands on both apps
@app.command("setup", requires_connection=True)
def setup(
    github_repository: str = typer.Option(
        ...,
        "--github-repository",
        help="GitHub repository in format 'owner/repo'",
        prompt="Enter GitHub repository (owner/repo)",
    ),
    **options,
):
    return _setup_command(github_repository=github_repository, **options)


@app.command("status", requires_connection=True)
def status(**options):
    return _status_command(**options)


@app.command("remove", requires_connection=True)
def remove(**options):
    return _remove_command(**options)
