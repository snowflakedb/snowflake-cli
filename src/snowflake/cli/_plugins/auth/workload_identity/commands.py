# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import typer
from snowflake.cli._app.auth.oidc_providers import (
    OidcProviderType,
)
from snowflake.cli._plugins.auth.workload_identity.manager import (
    WorkloadIdentityManager,
)
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.output.types import MessageResult

app = SnowTyperFactory(
    name="workload-identity",
    help="Manages GitHub workload identity federation authentication.",
)


@app.command("setup", requires_connection=True)
def setup(
    github_repository: str = typer.Option(
        ...,
        "--github-repository",
        show_default=False,
        help="GitHub repository in format 'owner/repo'",
        prompt="Enter GitHub repository (owner/repo)",
    ),
    **options,
):
    """
    Sets up GitHub workload identity federation for authentication.
    """
    result = WorkloadIdentityManager().setup(github_repository=github_repository)
    return MessageResult(result)


@app.command("read", requires_connection=False)
def read(
    _type: str = typer.Option(
        "auto",
        "--type",
        help=f"Type of read operation to perform (e.g., '{OidcProviderType.GITHUB.value}', 'auto')",
    ),
    **options,
):
    """
    Reads OIDC token based on the specified type.
    Use 'auto' to auto-detect available providers.
    """
    result = WorkloadIdentityManager().read(provider_type=_type)
    return MessageResult(result)
