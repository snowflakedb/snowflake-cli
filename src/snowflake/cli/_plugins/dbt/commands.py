# Copyright (c) 2025 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import logging
from pathlib import Path
from typing import Optional

import typer
from snowflake.cli._plugins.dbt.constants import DBT_COMMANDS
from snowflake.cli._plugins.dbt.manager import DBTManager
from snowflake.cli.api.commands.decorators import global_options_with_connection
from snowflake.cli.api.commands.flags import identifier_argument
from snowflake.cli.api.commands.snow_typer import SnowTyperFactory
from snowflake.cli.api.feature_flags import FeatureFlag
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.output.types import CommandResult, QueryResult

app = SnowTyperFactory(
    name="dbt",
    help="Manages dbt on Snowflake projects",
    is_hidden=FeatureFlag.ENABLE_DBT_POC.is_disabled,
)
log = logging.getLogger(__name__)


DBTNameArgument = identifier_argument(sf_object="DBT Object", example="my_pipeline")


@app.command(
    "list",
    requires_connection=True,
)
def list_dbts(
    **options,
) -> CommandResult:
    """
    List all dbt on Snowflake projects.
    """
    return QueryResult(DBTManager().list())


@app.command(
    "deploy",
    requires_connection=True,
)
def deploy_dbt(
    name: FQN = DBTNameArgument,
    source: Optional[str] = typer.Option(
        help="Path to directory containing dbt files to deploy. Defaults to current working directory.",
        show_default=False,
        default=None,
    ),
    force: Optional[bool] = typer.Option(
        False,
        help="Overwrites conflicting files in the project, if any.",
    ),
    dbt_version: Optional[str] = typer.Option(
        None,
        help="Version of dbt tool to be used. Taken from dbt_project.yml if not provided.",
    ),
    dbt_adapter_version: Optional[str] = typer.Option(
        None,
        help="dbt-snowflake adapter version to be used",
    ),
    execute_in_warehouse: Optional[str] = typer.Option(
        None, help="Warehouse to use when running `dbt execute` commands"
    ),
    **options,
) -> CommandResult:
    """
    Copy dbt files and create or update dbt on Snowflake project.
    """
    if source is None:
        path = Path.cwd()
    else:
        path = Path(source)
    return QueryResult(
        DBTManager().deploy(
            path.resolve(),
            name,
            dbt_version,
            dbt_adapter_version,
            execute_in_warehouse,
            force=force,
        )
    )


# `execute` is a pass through command group, meaning that all params after command should be passed over as they are,
# suppressing usual CLI behaviour for displaying help or formatting options.
dbt_execute_app = SnowTyperFactory(
    name="execute",
    help="Execute a dbt command on Snowflake",
)
app.add_typer(dbt_execute_app)


@dbt_execute_app.callback()
@global_options_with_connection
def before_callback(
    name: FQN = DBTNameArgument,
    **options,
):
    """Handles global options passed before the command and takes pipeline name to be accessed through child context later"""
    pass


for cmd in DBT_COMMANDS:

    @dbt_execute_app.command(
        name=cmd,
        requires_connection=False,
        requires_global_options=False,
        context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
        help=f"Execute {cmd} command on Snowflake.",
        add_help_option=False,
    )
    def _dbt_execute(
        ctx: typer.Context,
    ) -> CommandResult:
        dbt_cli_args = ctx.args
        dbt_command = ctx.command.name
        name = ctx.parent.params["name"]
        return QueryResult(DBTManager().execute(dbt_command, name, *dbt_cli_args))
