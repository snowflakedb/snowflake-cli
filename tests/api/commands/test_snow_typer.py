# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from functools import partial
from unittest import mock
from unittest.mock import MagicMock

import pytest
import typer
from snowflake.cli.api.commands.snow_typer import (
    PREVIEW_PREFIX,
    SnowTyper,
    SnowTyperFactory,
)
from snowflake.cli.api.output.types import MessageResult
from typer.testing import CliRunner


def class_factory(
    pre_execute=None,
    result_handler=None,
    exception_handler=None,
    post_execute=None,
):
    class _CustomTyper(SnowTyper):
        @staticmethod
        def pre_execute(execution, require_warehouse):
            if pre_execute:
                pre_execute(execution)

        @staticmethod
        def post_execute(execution):
            if post_execute:
                post_execute(execution)

        @staticmethod
        def process_result(result):
            if result_handler:
                result_handler(result)

        @staticmethod
        def exception_handler(err, execution):
            if exception_handler:
                exception_handler(err, execution)

        def create_instance(self):
            return self

    return _CustomTyper


_ENABLED_FLAG = False


def app_factory(typer_cls):
    app = typer_cls(name="snow")

    @app.command("simple_cmd", requires_global_options=False, requires_connection=False)
    def simple_cmd(name: str = typer.Argument()):
        return MessageResult(f"hello {name}")

    @app.command("fail_cmd", requires_global_options=False, requires_connection=False)
    def fail_cmd(name: str = typer.Argument()):
        raise Exception("err")

    @app.command(
        "cmd_with_global_options",
        requires_global_options=True,
        requires_connection=False,
    )
    def cmd_with_global_options(name: str = typer.Argument()):
        return MessageResult(f"hello {name}")

    @app.command("cmd_with_connection_options", requires_connection=True)
    def cmd_with_connection_options(name: str = typer.Argument()):
        return MessageResult(f"hello {name}")

    @app.command("switchable_cmd", is_enabled=lambda: _ENABLED_FLAG)
    def cmd_witch_enabled_switch():
        return MessageResult("Enabled")

    @app.command(
        "interrupted_cmd",
        requires_global_options=False,
        requires_connection=False,
    )
    def interrupted_cmd(name: str = typer.Argument()):
        raise KeyboardInterrupt("err")

    return app.create_instance()


@pytest.fixture
def cli():
    def mock_cli(app):
        return partial(CliRunner().invoke, app)

    return mock_cli


def test_no_callbacks(cli):
    result = cli(app_factory(class_factory()))(["simple_cmd", "Norman"])
    assert result.exit_code == 0, result.output
    assert result.output == ""


def test_result_callbacks(cli):
    result = cli(app_factory(class_factory(result_handler=lambda x: print(x.message))))(
        ["simple_cmd", "Norman"]
    )
    assert result.exit_code == 0, result.output
    assert result.output.strip() == "hello Norman"


def test_pre_callback_green_path(cli):
    pre_execute = MagicMock()
    post_execute = MagicMock()
    exception_callback = MagicMock()
    result = cli(
        app_factory(
            class_factory(
                pre_execute=pre_execute,
                post_execute=post_execute,
                exception_handler=exception_callback,
            )
        )
    )(["simple_cmd", "Norman"])
    assert result.exit_code == 0, result.output

    assert pre_execute.called
    assert post_execute.called
    assert not exception_callback.called

    # ensure duration metric captured is greater than or equal to 0
    assert post_execute.call_args.args[0].get_duration() >= 0


def test_pre_callback_error_path(cli):
    pre_execute = MagicMock()
    post_execute = MagicMock()
    exception_callback = MagicMock()
    result_handler = MagicMock()

    result = cli(
        app_factory(
            class_factory(
                pre_execute=pre_execute,
                post_execute=post_execute,
                exception_handler=exception_callback,
                result_handler=result_handler,
            )
        )
    )(["fail_cmd", "Norman"])
    assert result.exit_code == 1, result.output

    assert pre_execute.called
    assert post_execute.called
    assert not result_handler.called
    assert exception_callback.called
    assert len(exception_callback.call_args_list) == 1

    # ensure duration metric captured is greater than or equal to 0
    assert post_execute.call_args.args[0].get_duration() >= 0


def test_command_without_any_options(cli, os_agnostic_snapshot):
    result = cli(app_factory(SnowTyperFactory))(["simple_cmd", "--help"])
    assert result.output == os_agnostic_snapshot


def test_command_with_global_options(cli, os_agnostic_snapshot):
    result = cli(app_factory(SnowTyperFactory))(["cmd_with_global_options", "--help"])
    assert result.output == os_agnostic_snapshot


def test_command_with_connection_options(cli, os_agnostic_snapshot):
    result = cli(app_factory(SnowTyperFactory))(
        ["cmd_with_connection_options", "--help"]
    )
    assert result.output == os_agnostic_snapshot


def test_enabled_command_is_visible(cli, os_agnostic_snapshot):
    global _ENABLED_FLAG
    _ENABLED_FLAG = True
    result = cli(app_factory(SnowTyperFactory))(["switchable_cmd", "--help"])
    assert result.exit_code == 0
    assert result.output == os_agnostic_snapshot


def test_enabled_command_is_not_visible(cli, os_agnostic_snapshot):
    global _ENABLED_FLAG
    _ENABLED_FLAG = False
    result = cli(app_factory(SnowTyperFactory))(["switchable_cmd", "--help"])
    assert result.exit_code == 2
    assert result.output == os_agnostic_snapshot


@mock.patch("snowflake.cli._app.telemetry.log_command_usage")
def test_snow_typer_pre_execute_sends_telemetry(mock_log_command_usage, cli):
    result = cli(app_factory(SnowTyperFactory))(["simple_cmd", "Norma"])

    assert result.exit_code == 0
    mock_log_command_usage.assert_called_once_with(mock.ANY)


@mock.patch("snowflake.cli._app.telemetry.flush_telemetry")
def test_snow_typer_post_execute_sends_telemetry(mock_flush_telemetry, cli):
    result = cli(app_factory(SnowTyperFactory))(["simple_cmd", "Norma"])
    assert result.exit_code == 0
    mock_flush_telemetry.assert_called_once_with()


@mock.patch("snowflake.cli._app.printing.print_result")
def test_snow_typer_result_callback_sends_telemetry(mock_print_result, cli):
    result = cli(app_factory(SnowTyperFactory))(["simple_cmd", "Norma"])
    assert result.exit_code == 0
    assert mock_print_result.call_count == 1
    assert mock_print_result.call_args.args[0].message == "hello Norma"


def test_snow_typer_with_keyboard_interrupt(cli):
    pre_execute = MagicMock()
    post_execute = MagicMock()
    exception_callback = MagicMock()
    result_handler = MagicMock()

    result = cli(
        app_factory(
            class_factory(
                pre_execute=pre_execute,
                post_execute=post_execute,
                exception_handler=exception_callback,
                result_handler=result_handler,
            )
        )
    )(["interrupted_cmd", "Norman"])
    assert result.exit_code == 1, result.output

    assert pre_execute.called
    assert post_execute.called
    assert not result_handler.called
    assert exception_callback.called
    assert len(exception_callback.call_args_list) == 1

    # ensure duration metric captured is greater than or equal to 0
    assert post_execute.call_args.args[0].get_duration() >= 0


# Tests for the preview functionality
def test_snow_typer_factory_preview_adds_prefix_to_app_help():
    """Test that preview=True on SnowTyperFactory adds PREVIEW_PREFIX to app help text."""
    app = SnowTyperFactory(
        name="test_app",
        help="This is a test app.",
        preview=True,
    )

    instance = app.create_instance()
    assert instance.info.help == f"{PREVIEW_PREFIX}This is a test app."


def test_snow_typer_factory_preview_prevents_double_prefix():
    """Test that preview=True does not add double PREVIEW_PREFIX prefix."""
    app = SnowTyperFactory(
        name="test_app",
        help=f"{PREVIEW_PREFIX}This is already prefixed.",
        preview=True,
    )

    instance = app.create_instance()
    assert instance.info.help == f"{PREVIEW_PREFIX}This is already prefixed."


def test_snow_typer_factory_preview_propagates_to_commands():
    """Test that preview=True on SnowTyperFactory propagates to individual commands."""
    app = SnowTyperFactory(
        name="test_app",
        help="This is a test app.",
        preview=True,
    )

    @app.command("test_cmd", requires_global_options=False, requires_connection=False)
    def test_cmd():
        """This is a test command."""
        return MessageResult("test")

    instance = app.create_instance()

    # Get the command and check its help text
    registered_cmds = instance.registered_commands
    assert len(registered_cmds) == 1
    cmd_info = registered_cmds[0]
    assert cmd_info.callback.__doc__ == f"{PREVIEW_PREFIX}This is a test command."


def test_snow_typer_individual_command_preview():
    """Test that individual commands can have preview=True."""
    app = SnowTyperFactory(
        name="test_app",
        help="This is a test app.",
        preview=False,  # App itself is not preview
    )

    @app.command(
        "preview_cmd",
        requires_global_options=False,
        requires_connection=False,
        preview=True,
    )
    def preview_cmd():
        """This command is in preview."""
        return MessageResult("test")

    @app.command("normal_cmd", requires_global_options=False, requires_connection=False)
    def normal_cmd():
        """This command is normal."""
        return MessageResult("test")

    instance = app.create_instance()

    # Check that only the preview command has the prefix
    registered_cmds = instance.registered_commands
    assert len(registered_cmds) == 2

    # Find commands by name
    preview_cmd_info = None
    normal_cmd_info = None
    for cmd_info in registered_cmds:
        if cmd_info.name == "preview_cmd":
            preview_cmd_info = cmd_info
        elif cmd_info.name == "normal_cmd":
            normal_cmd_info = cmd_info

    assert preview_cmd_info is not None
    assert normal_cmd_info is not None
    assert (
        preview_cmd_info.callback.__doc__
        == f"{PREVIEW_PREFIX}This command is in preview."
    )
    assert normal_cmd_info.callback.__doc__ == "This command is normal."


def test_snow_typer_preview_works_with_help_parameter():
    """Test that preview=True works with help parameter in addition to docstrings."""
    app = SnowTyperFactory(
        name="test_app",
        help="This is a test app.",
        preview=False,  # App itself is not preview
    )

    @app.command(
        "help_param_cmd",
        requires_global_options=False,
        requires_connection=False,
        preview=True,
        help="This command uses help parameter.",
    )
    def help_param_cmd():
        return MessageResult("test")

    @app.command(
        "docstring_cmd",
        requires_global_options=False,
        requires_connection=False,
        preview=True,
    )
    def docstring_cmd():
        """This command uses docstring."""
        return MessageResult("test")

    instance = app.create_instance()

    # Check both commands have preview prefix
    registered_cmds = instance.registered_commands
    assert len(registered_cmds) == 2

    # Find commands by name
    help_param_cmd_info = None
    docstring_cmd_info = None
    for cmd_info in registered_cmds:
        if cmd_info.name == "help_param_cmd":
            help_param_cmd_info = cmd_info
        elif cmd_info.name == "docstring_cmd":
            docstring_cmd_info = cmd_info

    assert help_param_cmd_info is not None
    assert docstring_cmd_info is not None

    # The help parameter command should have PREVIEW_PREFIX in help
    assert (
        help_param_cmd_info.help == f"{PREVIEW_PREFIX}This command uses help parameter."
    )
    # The docstring command should have PREVIEW_PREFIX in docstring
    assert (
        docstring_cmd_info.callback.__doc__
        == f"{PREVIEW_PREFIX}This command uses docstring."
    )
