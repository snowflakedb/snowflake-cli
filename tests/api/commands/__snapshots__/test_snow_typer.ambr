# serializer version: 1
# name: test_command_with_connection_options
  '''
                                                                                  
   Usage: snow cmd_with_connection_options [OPTIONS] NAME                         
                                                                                  
  +- Arguments ------------------------------------------------------------------+
  | *    name      TEXT  [default: None] [required]                              |
  +------------------------------------------------------------------------------+
  +- Options --------------------------------------------------------------------+
  | --help  -h        Show this message and exit.                                |
  +------------------------------------------------------------------------------+
  +- Connection configuration ---------------------------------------------------+
  | --connection,--environment     -c      TEXT     Name of the connection, as   |
  |                                                 defined in your config.toml  |
  |                                                 file. Default: default.      |
  | --host                                 TEXT     Host address for the         |
  |                                                 connection. Overrides the    |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --port                                 INTEGER  Port for the connection.     |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --account,--accountname                TEXT     Name assigned to your        |
  |                                                 Snowflake account. Overrides |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --user,--username                      TEXT     Username to connect to       |
  |                                                 Snowflake. Overrides the     |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --password                             TEXT     Snowflake password.          |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --authenticator                        TEXT     Snowflake authenticator.     |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --private-key-file,--privateâ€¦          TEXT     Snowflake private key file   |
  |                                                 path. Overrides the value    |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --token-file-path                      TEXT     Path to file with an OAuth   |
  |                                                 token that should be used    |
  |                                                 when connecting to Snowflake |
  | --database,--dbname                    TEXT     Database to use. Overrides   |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --schema,--schemaname                  TEXT     Database schema to use.      |
  |                                                 Overrides the value          |
  |                                                 specified for the            |
  |                                                 connection.                  |
  | --role,--rolename                      TEXT     Role to use. Overrides the   |
  |                                                 value specified for the      |
  |                                                 connection.                  |
  | --warehouse                            TEXT     Warehouse to use. Overrides  |
  |                                                 the value specified for the  |
  |                                                 connection.                  |
  | --temporary-connection         -x               Uses a connection defined    |
  |                                                 with command line            |
  |                                                 parameters, instead of one   |
  |                                                 defined in config            |
  | --mfa-passcode                         TEXT     Token to use for             |
  |                                                 multi-factor authentication  |
  |                                                 (MFA)                        |
  | --enable-diag                                   Whether to generate a        |
  |                                                 connection diagnostic        |
  |                                                 report.                      |
  | --diag-log-path                        TEXT     Path for the generated       |
  |                                                 report. Defaults to system   |
  |                                                 temporary directory.         |
  | --diag-allowlist-path                  TEXT     Path to a JSON file          |
  |                                                 containing allowlist         |
  |                                                 parameters.                  |
  +------------------------------------------------------------------------------+
  +- Global configuration -------------------------------------------------------+
  | --format           [TABLE|JSON]  Specifies the output format.                |
  |                                  [default: TABLE]                            |
  | --verbose  -v                    Displays log entries for log levels info    |
  |                                  and higher.                                 |
  | --debug                          Displays log entries for log levels debug   |
  |                                  and higher; debug logs contain additional   |
  |                                  information.                                |
  | --silent                         Turns off intermediate output to console.   |
  +------------------------------------------------------------------------------+
  
  
  '''
# ---
# name: test_command_with_global_options
  '''
                                                                                  
   Usage: snow cmd_with_global_options [OPTIONS] NAME                             
                                                                                  
  +- Arguments ------------------------------------------------------------------+
  | *    name      TEXT  [default: None] [required]                              |
  +------------------------------------------------------------------------------+
  +- Options --------------------------------------------------------------------+
  | --help  -h        Show this message and exit.                                |
  +------------------------------------------------------------------------------+
  +- Global configuration -------------------------------------------------------+
  | --format           [TABLE|JSON]  Specifies the output format.                |
  |                                  [default: TABLE]                            |
  | --verbose  -v                    Displays log entries for log levels info    |
  |                                  and higher.                                 |
  | --debug                          Displays log entries for log levels debug   |
  |                                  and higher; debug logs contain additional   |
  |                                  information.                                |
  | --silent                         Turns off intermediate output to console.   |
  +------------------------------------------------------------------------------+
  
  
  '''
# ---
# name: test_command_without_any_options
  '''
                                                                                  
   Usage: snow simple_cmd [OPTIONS] NAME                                          
                                                                                  
  +- Arguments ------------------------------------------------------------------+
  | *    name      TEXT  [default: None] [required]                              |
  +------------------------------------------------------------------------------+
  +- Options --------------------------------------------------------------------+
  | --help  -h        Show this message and exit.                                |
  +------------------------------------------------------------------------------+
  
  
  '''
# ---
# name: test_enabled_command_is_not_visible
  '''
  Usage: snow [OPTIONS] COMMAND [ARGS]...
  Try 'snow --help' for help.
  +- Error ----------------------------------------------------------------------+
  | No such command 'switchable_cmd'.                                            |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_enabled_command_is_visible
  '''
                                                                                  
   Usage: snow switchable_cmd [OPTIONS]                                           
                                                                                  
  +- Options --------------------------------------------------------------------+
  | --help  -h        Show this message and exit.                                |
  +------------------------------------------------------------------------------+
  +- Global configuration -------------------------------------------------------+
  | --format           [TABLE|JSON]  Specifies the output format.                |
  |                                  [default: TABLE]                            |
  | --verbose  -v                    Displays log entries for log levels info    |
  |                                  and higher.                                 |
  | --debug                          Displays log entries for log levels debug   |
  |                                  and higher; debug logs contain additional   |
  |                                  information.                                |
  | --silent                         Turns off intermediate output to console.   |
  +------------------------------------------------------------------------------+
  
  
  '''
# ---
